draft-barnes-acme.md

```diff
@@ -516,18 +544,234 @@ Link: <https://example.com/acme/terms>;rel="terms-of-service"
+
+### Recovery Keys
+
+If the client wishes to establish a secret key with the server that it can use to recover this account later (a "recovery key"), then it must perform a simple key agreement protocol as part of the new-registration transaction.  The client and server perform an ECDH exchange through the new-registration transaction, and the result is the recovery key.
+
+To request a recovery key, the client includes a "recoveryKey" field in its new-registration request.  The value of this field is a JSON object.
+
+client (required, JWK):
+: The client's ECDH public key
 
+server (optional, JWK):
+: The server's ECDH public key
 
-### Account Recovery
+length (optional, number):
+: The length, in bits, of the derived secret.  The value MUST be a multiple of 8.
+
+confirm (optional, string):
+: The confirmation value, base64-encoded
+
+In the client's request, this object contains a JWK for a random ECDH public key generated by the client.
+
+~~~~~~~~~~
+
+POST /acme/new-registration HTTP/1.1
+Host: example.com
+
+{
+  "resource": "new-reg",
+  "contact": [
+    "mailto:cert-admin@example.com",
+    "tel:+12025551212"
+  ],
+  "recoveryKey": {
+    "client": { 
+      "kty": "EC", 
+      "crv": "P-256",
+      "x": "...","y": "..." 
+    }
+  }
+}
+/* Signed as JWS */
+
+~~~~~~~~~~
+
+If the server agrees to create a recovery key, then it generates its own random ECDH key pair and combines it with with the client's public key as described in {{key-agreement}} above.  The derived secret value is the recovery key.  The server chooses the length of the recovery key and indicates this in the "length" field of the recoveryKey object (in units of bits).  The server MUST generate a confirmation value and include it in the "confirm" field of the recoveryKey object (base64-encoded).
+
+
+~~~~~~~~~~
+
+HTTP/1.1 201 Created
+Content-Type: application/json
+Location: https://example.com/acme/reg/asdf
+
+{
+  "key": { /* JWK from JWS header */ },
+
+  "contact": [
+    "mailto:cert-admin@example.com",
+    "tel:+12025551212"
+  ],
+
+  "recoveryKey": {
+    "client": {
+      "kty": "EC",
+      "crv": "P-256",
+      "x": "...","y": "..."
+    },
+    "server": {
+      "kty": "EC",
+      "crv": "P-256",
+      "x": "...","y": "..."
+    },
+    "length": 80,
+    "confirm": "_VPtXjAFDBC0O_JTnv8MCS7uTyb1NVlQtUvr5xcLjOo"
+  }
+}
+
+~~~~~~~~~~
+
+On receiving the server's response, the client MUST compute the recovery key and check that it matches the confirmation value.  (This ensures that the "length" field was not tampered with in transit.)  If the confirmation value does not match, then the client should discard the recovery key; it will not be usable for recovery on this registration.
+
```

Comment:
Except that it doesn't demonstrate that the server's PK wasn't modified in transit. How are you in a worse situatio here?